File: release-artifacts/tests/fixtures/sso-release-origin.py
#!/usr/bin/env python3
"""Serial, owned three-Origin fixture; never builds/stages a deployable release.

Uses production Nginx templates and freshly built Apps with the existing Java
SSO/Redis/MySQL fixture. System identity/menu and business token minting remain
explicit test doubles. Only recorded container/network IDs are removed.
"""
import argparse
import hashlib
import json
import os
from pathlib import Path
import re
import signal
import socket
import subprocess
import tempfile
import time
import uuid

ROOT = Path(__file__).resolve().parents[3]
IMAGES = {
    'nginx': 'nginx:1.31.1@sha256:608a100c71651bf5b773c89083b4a1ad7ef4b2bd05d7a7e552271e03123692ad',
    'mysql': 'mysql:8.4.9@sha256:c36050afdca850f23cef85703f84c7531a5ae155a11b5ee1c60acb09937c4084',
    'redis': 'redis:7.4-alpine@sha256:ff02b58f971e7d7d156a1267e283fcbbeee91773b6aa36c49dac28ecfe28eadf',
}


def output(command):
    return subprocess.check_output(command, text=True, stderr=subprocess.STDOUT).strip()


def inventory():
    return {kind: sorted(output(['docker', *command]).splitlines()) for kind, command in {
        'containers': ['ps', '-aq', '--no-trunc'], 'networks': ['network', 'ls', '-q', '--no-trunc'],
        'volumes': ['volume', 'ls', '-q'],
    }.items()}


def available_port(host='127.0.0.1'):
    with socket.socket() as listener:
        listener.bind((host, 0))
        return listener.getsockname()[1]


def main():
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--evidence', type=Path, required=True, help='New result JSON; logs share its filename stem')
    args = parser.parse_args()
    evidence = args.evidence.resolve()
    if evidence.exists():
        raise SystemExit('Refusing to overwrite prior evidence')
    evidence.parent.mkdir(parents=True, exist_ok=True)
    owner = 'namewta-sso-release-' + uuid.uuid4().hex[:12]
    record = {'owner': owner, 'images': IMAGES, 'commands': [], 'containers': [], 'network': None,
              'cleanup': [], 'before': inventory(), 'boundary': 'production Nginx/App/SSO/Redis/MySQL; mocked System identity/menu/token minting'}

    def save():
        evidence.write_text(json.dumps(record, indent=2) + '\n')

    def run(command, cwd=ROOT, env=None, timeout=900):
        log = evidence.with_name(evidence.stem + '-' + str(len(record['commands']) + 1) + '.log')
        started = time.time()
        with log.open('w') as stream:
            process = subprocess.Popen(command, cwd=cwd, env=env, stdout=stream, stderr=subprocess.STDOUT, start_new_session=True)
            try:
                code = process.wait(timeout=timeout)
            except BaseException:
                os.killpg(process.pid, signal.SIGTERM)
                try:
                    process.wait(timeout=5)
                except subprocess.TimeoutExpired:
                    os.killpg(process.pid, signal.SIGKILL)
                    process.wait()
                raise
        # Never expose transient OAuth query values through fixture error reports.
        log.write_text(re.sub(r'''([?&](?:code|state|code_challenge|code_verifier)=)[^&\s"'<>]+''', r'\1[REDACTED]', log.read_text()))
        record['commands'].append({'command': command, 'cwd': str(cwd), 'exit_code': code,
                                   'seconds': round(time.time() - started, 2), 'log': log.name})
        save()
        print(f'{command[0]} {command[1]}: exit={code}, log={log.name}', flush=True)
        if code:
            raise RuntimeError('Fixture command failed; inspect recorded log')

    def container(name, arguments):
        identifier = output(['docker', 'create', '--name', owner + '-' + name, '--label', 'namewta.test.owner=' + owner, *arguments])
        record['containers'].append({'name': name, 'id': identifier})
        save()
        output(['docker', 'start', identifier])
        return identifier

    def ready(command, expected, attempts=90):
        for _ in range(attempts):
            result = subprocess.run(command, capture_output=True, text=True)
            if result.returncode == 0 and expected in result.stdout:
                return
            time.sleep(1)
        raise RuntimeError('Owned service readiness timeout')

    exit_code = 1
    try:
        with tempfile.TemporaryDirectory(prefix=owner + '-') as temporary:
            work = Path(temporary)
            shim = work / 'pnpm'
            shim.write_text('#!/bin/sh\nexec corepack pnpm "$@"\n')
            shim.chmod(0o700)
            environment = {**os.environ, 'PATH': str(work) + os.pathsep + os.environ['PATH'],
                           'npm_config_workspace_concurrency': '1', 'RAYON_NUM_THREADS': '1'}
            run(['corepack', 'pnpm', 'build:dependencies'], ROOT / 'frontend', environment)
            for app, prefix in [('admin-web', 'admin-app'), ('home-web', 'home-app'), ('sso-web', 'sso-app')]:
                build_environment = {**environment, 'VITE_APP_CONTEXT_PATH': '/' + prefix + '/',
                                     'VITE_APP_BASE_API': '/' + prefix + '/prod-api', 'VITE_SSO_API': ''}
                run(['corepack', 'pnpm', '--filter', '@namewta/' + app, 'build:prod'], ROOT / 'frontend', build_environment)
                dist = ROOT / 'frontend/apps' / app / 'dist'
                assert json.loads((dist / 'build-mode.json').read_text()) == {'app': app, 'mode': 'production'}
                record.setdefault('built_apps', {})[app] = hashlib.sha256((dist / 'index.html').read_bytes()).hexdigest()

            network = output(['docker', 'network', 'create', '--label', 'namewta.test.owner=' + owner, owner])
            record['network'] = network
            save()
            gateway = json.loads(output(['docker', 'network', 'inspect', network]))[0]['IPAM']['Config'][0]['Gateway']
            backend_port = available_port(gateway)
            ports = {app: available_port() for app in ['admin', 'home', 'sso']}
            origins = {app: f'https://{"localhost" if app == "sso" else "127.0.0.1"}:{port}' for app, port in ports.items()}
            record['origins'] = origins
            cert = work / 'cert'
            cert.mkdir()
            run(['openssl', 'req', '-x509', '-newkey', 'rsa:2048', '-nodes', '-days', '1', '-subj', '/CN=localhost',
                 '-addext', 'subjectAltName=DNS:localhost,IP:127.0.0.1', '-keyout', str(cert / 'privkey.pem'), '-out', str(cert / 'fullchain.pem')])
            templates = ROOT / 'release-artifacts/docker/frontend/nginx'
            backend = f'{gateway}:{backend_port}'

            def nginx(name, template, html, bindings, port=None, alias=None, tls_dir=None):
                arguments = ['--network', network, '-v', f'{template}:/etc/nginx/templates/default.conf.template:ro',
                             '-v', f'{html}:/usr/share/nginx/html:ro']
                if alias:
                    arguments += ['--network-alias', alias]
                if port:
                    arguments += ['-p', f'127.0.0.1:{port}:443']
                if tls_dir:
                    arguments += ['-v', f'{cert}:/etc/nginx/cert/{tls_dir}:ro']
                for key, value in bindings.items():
                    arguments += ['-e', key + '=' + value]
                for host in ['namewta-monitor-admin', 'namewta-snailjob-server', 'namewta-snailai-server', 'nacos']:
                    arguments += ['--add-host', host + ':' + gateway]
                identifier = container(name, [*arguments, IMAGES['nginx']])
                record.setdefault('templates', {})[str(template.relative_to(ROOT))] = hashlib.sha256(template.read_bytes()).hexdigest()
                ready(['docker', 'exec', identifier, 'nginx', '-t'], '', attempts=15)
                return identifier

            for app in ['admin', 'home']:
                nginx(app, templates / f'apps/nginx-{app}-web.conf.template', ROOT / f'frontend/apps/{app}-web/dist',
                      {'APP_PREFIX': app + '-app', 'BACKEND_SERVER1': backend, 'BACKEND_SERVER2': backend},
                      alias=f'namewta-nginx-{app}-web')
            for app in ['admin', 'home']:
                nginx(app + '-tls', templates / 'lb/nginx-lb-tls.conf.template', templates / 'html',
                      {'APP_ADMIN_WEB_PREFIX': 'admin-app', 'APP_HOME_WEB_PREFIX': 'home-app', 'LB_SERVER_NAME': '127.0.0.1'},
                      port=ports[app], tls_dir='lb')
            nginx('sso-tls', templates / 'apps/nginx-sso-web-tls.conf.template', ROOT / 'frontend/apps/sso-web/dist',
                  {'APP_PREFIX': 'sso-app', 'APP_ORIGIN': origins['sso'], 'BACKEND_SERVER1': backend, 'BACKEND_SERVER2': backend},
                  port=ports['sso'], tls_dir='sso-web')

            database = 'namewta_sso_test_' + uuid.uuid4().hex[:12]
            mysql = container('mysql', ['-p', '127.0.0.1::3306', '-e', 'MYSQL_ROOT_PASSWORD=owned-sso-test-only',
                              '-e', 'MYSQL_DATABASE=' + database, IMAGES['mysql'], '--character-set-server=utf8mb4', '--collation-server=utf8mb4_general_ci'])
            ready(['docker', 'exec', mysql, 'mysqladmin', 'ping', '-h', '127.0.0.1', '-uroot', '-powned-sso-test-only', '--silent'], 'alive')
            mysql_port = output(['docker', 'port', mysql, '3306/tcp']).rsplit(':', 1)[1]
            redis = container('redis', ['-p', '127.0.0.1::6379', IMAGES['redis'], 'redis-server', '--save', '', '--appendonly', 'no'])
            ready(['docker', 'exec', redis, 'redis-cli', 'ping'], 'PONG')
            redis_port = output(['docker', 'port', redis, '6379/tcp']).rsplit(':', 1)[1]
            command = ['./mvnw', '-B', '-ntp', '-pl', 'wta-admin', '-am', 'test', '-Dtest=SsoHttpsSessionIntegrationTest,AuthClientContextSsoUnitTest',
                       '-Dsurefire.failIfNoSpecifiedTests=false', '-Dsso.release.integration=true',
                       '-Dsso.release.backend.host=' + gateway, '-Dsso.release.backend.port=' + str(backend_port),
                       '-Dsso.redis.integration.port=' + redis_port,
                       '-Dsso.mysql.integration.url=jdbc:mysql://127.0.0.1:' + mysql_port + '/' + database + '?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=Asia/Shanghai',
                       '-Dnamewta.repo.root=' + str(ROOT), '-Dnamewta.sql.root=' + str(ROOT / 'release-artifacts/docker/infrastructure/mysql/init'),
                       *['-Dsso.release.' + app + '.origin=' + origin for app, origin in origins.items()]]
            environment['MAVEN_OPTS'] = environment.get('MAVEN_OPTS', '') + ' -Daether.connector.requestTimeout=60000 -Daether.connector.connectTimeout=10000 -Daether.connector.basic.threads=1 -Dmaven.artifact.threads=1'
            run(command, ROOT / 'backend', environment)
            report = ROOT / 'backend/wta-admin/target/surefire-reports/TEST-org.namewta.test.sso.SsoHttpsSessionIntegrationTest.xml'
            import xml.etree.ElementTree as xml
            suite = xml.parse(report).getroot().attrib
            record['tests'] = {key: int(suite[key]) for key in ['tests', 'failures', 'errors', 'skipped']}
            assert record['tests'] == {'tests': 1, 'failures': 0, 'errors': 0, 'skipped': 0}
            context_report = ROOT / 'backend/wta-admin/target/surefire-reports/TEST-org.namewta.web.controller.AuthClientContextSsoUnitTest.xml'
            context_suite = xml.parse(context_report).getroot().attrib
            record['client_context_tests'] = {key: int(context_suite[key]) for key in ['tests', 'failures', 'errors', 'skipped']}
            assert record['client_context_tests'] == {'tests': 3, 'failures': 0, 'errors': 0, 'skipped': 0}
            exit_code = 0
    except Exception as failure:
        record['failure'] = str(failure)
        print(str(failure), flush=True)
    finally:
        for row in reversed(record['containers']):
            if row['name'] not in ('mysql', 'redis'):
                logs = subprocess.run(['docker', 'logs', row['id']], capture_output=True, text=True)
                safe_logs = re.sub(r'''\?[^\s"']+''', '?[REDACTED]', logs.stdout + logs.stderr)
                evidence.with_name(evidence.stem + '-' + row['name'] + '.log').write_text(safe_logs)
            removed = subprocess.run(['docker', 'rm', '-fv', row['id']], capture_output=True, text=True)
            record['cleanup'].append({'id': row['id'], 'exit_code': removed.returncode})
            if removed.returncode:
                exit_code = 1
        if record['network']:
            removed = subprocess.run(['docker', 'network', 'rm', record['network']], capture_output=True, text=True)
            record['cleanup'].append({'network': record['network'], 'exit_code': removed.returncode})
            if removed.returncode:
                exit_code = 1
        record['after'] = inventory()
        record['resources_restored'] = record['before'] == record['after']
        if not record['resources_restored']:
            exit_code = 1
        record['exit_code'] = exit_code
        save()
    return exit_code


if __name__ == '__main__':
    raise SystemExit(main())

File: release-artifacts/tests/fixtures/sso-release-origin.mjs
#!/usr/bin/env node
// Serial, owned three-Origin fixture; never builds/stages a deployable release.
//
// Uses production Nginx templates and freshly built Apps with the existing Java
// SSO/Redis/MySQL fixture. System identity/menu and business token minting remain
// explicit test doubles. Only recorded container/network IDs are removed.

import { spawn, spawnSync } from 'node:child_process';
import crypto from 'node:crypto';
import fs from 'node:fs';
import net from 'node:net';
import os from 'node:os';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = path.dirname(path.dirname(path.dirname(path.dirname(fileURLToPath(import.meta.url)))));
const IMAGES = {
  nginx: 'nginx:1.31.1@sha256:608a100c71651bf5b773c89083b4a1ad7ef4b2bd05d7a7e552271e03123692ad',
  mysql: 'mysql:8.4.9@sha256:c36050afdca850f23cef85703f84c7531a5ae155a11b5ee1c60acb09937c4084',
  redis: 'redis:7.4-alpine@sha256:ff02b58f971e7d7d156a1267e283fcbbeee91773b6aa36c49dac28ecfe28eadf',
};

function output(command) {
  const result = spawnSync(command[0], command.slice(1), { encoding: 'utf8' });
  if (result.error) throw result.error;
  const combined = `${result.stdout ?? ''}${result.stderr ?? ''}`;
  if (result.status !== 0) throw new Error(combined.trim() || `command failed (${result.status})`);
  return (result.stdout ?? '').trim();
}

function inventory() {
  const queries = {
    containers: ['ps', '-aq', '--no-trunc'],
    networks: ['network', 'ls', '-q', '--no-trunc'],
    volumes: ['volume', 'ls', '-q'],
  };
  return Object.fromEntries(Object.entries(queries).map(([kind, command]) => [kind, output(['docker', ...command]).split('\n').filter(Boolean).sort()]));
}

function availablePort(host = '127.0.0.1') {
  return new Promise((resolve, reject) => {
    const server = net.createServer();
    server.once('error', reject);
    server.listen(0, host, () => {
      const { port } = server.address();
      server.close(() => resolve(port));
    });
  });
}

function sleep(milliseconds) {
  return new Promise((resolve) => setTimeout(resolve, milliseconds));
}

function killGroup(pid, signalName) {
  try { process.kill(-pid, signalName); }
  catch (error) { if (error.code !== 'ESRCH') throw error; }
}

function assertSame(actual, expected) {
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`${JSON.stringify(actual)} != ${JSON.stringify(expected)}`);
  }
}

function usage(message) {
  const error = new Error(message);
  error.usage = true;
  throw error;
}

function parseArgs(argv) {
  let evidence = null;
  for (let index = 0; index < argv.length; index += 1) {
    const token = argv[index];
    if (token === '--evidence') {
      evidence = argv[index + 1];
      if (!evidence || evidence.startsWith('--')) usage('missing value for --evidence');
      index += 1;
    } else usage(`unrecognized arguments: ${token}`);
  }
  if (!evidence) usage('missing required argument --evidence');
  return path.resolve(evidence);
}

async function main() {
  const evidence = parseArgs(process.argv.slice(2));
  if (fs.existsSync(evidence)) {
    console.error('Refusing to overwrite prior evidence');
    return 1;
  }
  fs.mkdirSync(path.dirname(evidence), { recursive: true });
  const owner = `namewta-sso-release-${crypto.randomBytes(6).toString('hex')}`;
  const record = {
    owner, images: IMAGES, commands: [], containers: [], network: null, cleanup: [], before: inventory(),
    boundary: 'production Nginx/App/SSO/Redis/MySQL; mocked System identity/menu/token minting',
  };
  const save = () => fs.writeFileSync(evidence, `${JSON.stringify(record, null, 2)}\n`);
  const stem = path.basename(evidence, path.extname(evidence));
  const sibling = (suffix) => path.join(path.dirname(evidence), `${stem}-${suffix}`);

  async function run(command, cwd = ROOT, env = null, timeout = 900) {
    const log = sibling(`${record.commands.length + 1}.log`);
    const started = Date.now();
    const fd = fs.openSync(log, 'w');
    const child = spawn(command[0], command.slice(1), {
      cwd: String(cwd), env: env ?? undefined, detached: true, stdio: ['ignore', fd, fd],
    });
    let timer;
    let interrupted = null;
    const finished = new Promise((resolve, reject) => {
      child.once('error', reject);
      child.once('exit', (code) => resolve(code ?? 1));
    });
    const stop = (error) => {
      interrupted = error;
      killGroup(child.pid, 'SIGTERM');
      const killer = setTimeout(() => killGroup(child.pid, 'SIGKILL'), 5000);
      finished.finally(() => clearTimeout(killer));
    };
    const previous = run.interrupt;
    run.interrupt = stop;
    let code;
    try {
      code = await Promise.race([
        finished,
        new Promise((_, reject) => { timer = setTimeout(() => reject(Object.assign(new Error('timeout'), { timeout: true })), timeout * 1000); }),
      ]);
      if (interrupted) throw interrupted;
    } catch (error) {
      if (!interrupted) stop(error);
      await finished.catch(() => {});
      throw error;
    } finally {
      clearTimeout(timer);
      run.interrupt = previous;
      fs.closeSync(fd);
    }
    const redacted = fs.readFileSync(log, 'utf8').replace(/([?&](?:code|state|code_challenge|code_verifier)=)[^&\s"'<>]+/g, '$1[REDACTED]');
    fs.writeFileSync(log, redacted);
    record.commands.push({
      command, cwd: String(cwd), exit_code: code, seconds: Math.round((Date.now() - started) / 10) / 100, log: path.basename(log),
    });
    save();
    console.log(`${command[0]} ${command[1]}: exit=${code}, log=${path.basename(log)}`);
    if (code) throw new Error('Fixture command failed; inspect recorded log');
  }

  function container(name, arguments_) {
    const identifier = output(['docker', 'create', '--name', `${owner}-${name}`, '--label', `namewta.test.owner=${owner}`, ...arguments_]);
    record.containers.push({ name, id: identifier });
    save();
    output(['docker', 'start', identifier]);
    return identifier;
  }

  async function ready(command, expected, attempts = 90) {
    for (let attempt = 0; attempt < attempts; attempt += 1) {
      const result = spawnSync(command[0], command.slice(1), { encoding: 'utf8' });
      if (result.status === 0 && (result.stdout ?? '').includes(expected)) return;
      await sleep(1000);
    }
    throw new Error('Owned service readiness timeout');
  }

  let exitCode = 1;
  let temporary = null;
  try {
    temporary = fs.mkdtempSync(path.join(os.tmpdir(), `${owner}-`));
    const shim = path.join(temporary, 'pnpm');
    fs.writeFileSync(shim, '#!/bin/sh\nexec corepack pnpm "$@"\n', { mode: 0o700 });
    const environment = {
      ...process.env,
      PATH: `${temporary}${path.delimiter}${process.env.PATH}`,
      npm_config_workspace_concurrency: '1',
      RAYON_NUM_THREADS: '1',
    };
    await run(['corepack', 'pnpm', 'build:dependencies'], path.join(ROOT, 'frontend'), environment);
    for (const [app, prefix] of [['admin-web', 'admin-app'], ['home-web', 'home-app'], ['sso-web', 'sso-app']]) {
      const buildEnvironment = {
        ...environment,
        VITE_APP_CONTEXT_PATH: `/${prefix}/`,
        VITE_APP_BASE_API: `/${prefix}/prod-api`,
        VITE_SSO_API: '',
      };
      await run(['corepack', 'pnpm', '--filter', `@namewta/${app}`, 'build:prod'], path.join(ROOT, 'frontend'), buildEnvironment);
      const dist = path.join(ROOT, 'frontend/apps', app, 'dist');
      assertSame(JSON.parse(fs.readFileSync(path.join(dist, 'build-mode.json'), 'utf8')), { app, mode: 'production' });
      record.built_apps ??= {};
      record.built_apps[app] = crypto.createHash('sha256').update(fs.readFileSync(path.join(dist, 'index.html'))).digest('hex');
    }

    const network = output(['docker', 'network', 'create', '--label', `namewta.test.owner=${owner}`, owner]);
    record.network = network;
    save();
    const gateway = JSON.parse(output(['docker', 'network', 'inspect', network]))[0].IPAM.Config[0].Gateway;
    const backendPort = await availablePort(gateway);
    const ports = {
      admin: await availablePort(),
      home: await availablePort(),
      sso: await availablePort(),
    };
    const origins = Object.fromEntries(Object.entries(ports).map(([app, port]) => [app, `https://${app === 'sso' ? 'localhost' : '127.0.0.1'}:${port}`]));
    record.origins = origins;
    const cert = path.join(temporary, 'cert');
    fs.mkdirSync(cert);
    await run(['openssl', 'req', '-x509', '-newkey', 'rsa:2048', '-nodes', '-days', '1', '-subj', '/CN=localhost',
      '-addext', 'subjectAltName=DNS:localhost,IP:127.0.0.1', '-keyout', path.join(cert, 'privkey.pem'), '-out', path.join(cert, 'fullchain.pem')]);
    const templates = path.join(ROOT, 'release-artifacts/docker/frontend/nginx');
    const backend = `${gateway}:${backendPort}`;

    async function nginx(name, template, html, bindings, { port = null, alias = null, tlsDir = null } = {}) {
      const arguments_ = ['--network', network, '-v', `${template}:/etc/nginx/templates/default.conf.template:ro`, '-v', `${html}:/usr/share/nginx/html:ro`];
      if (alias) arguments_.push('--network-alias', alias);
      if (port) arguments_.push('-p', `127.0.0.1:${port}:443`);
      if (tlsDir) arguments_.push('-v', `${cert}:/etc/nginx/cert/${tlsDir}:ro`);
      for (const [key, value] of Object.entries(bindings)) arguments_.push('-e', `${key}=${value}`);
      for (const host of ['namewta-monitor-admin', 'namewta-snailjob-server', 'nacos']) arguments_.push('--add-host', `${host}:${gateway}`);
      const identifier = container(name, [...arguments_, IMAGES.nginx]);
      record.templates ??= {};
      record.templates[path.relative(ROOT, template)] = crypto.createHash('sha256').update(fs.readFileSync(template)).digest('hex');
      await ready(['docker', 'exec', identifier, 'nginx', '-t'], '', 15);
      return identifier;
    }

    for (const app of ['admin', 'home']) {
      await nginx(app, path.join(templates, `apps/nginx-${app}-web.conf.template`), path.join(ROOT, `frontend/apps/${app}-web/dist`),
        { APP_PREFIX: `${app}-app`, BACKEND_SERVER1: backend, BACKEND_SERVER2: backend }, { alias: `namewta-nginx-${app}-web` });
    }
    for (const app of ['admin', 'home']) {
      await nginx(`${app}-tls`, path.join(templates, 'lb/nginx-lb-tls.conf.template'), path.join(templates, 'html'),
        { APP_ADMIN_WEB_PREFIX: 'admin-app', APP_HOME_WEB_PREFIX: 'home-app', LB_SERVER_NAME: '127.0.0.1' },
        { port: ports[app], tlsDir: 'lb' });
    }
    await nginx('sso-tls', path.join(templates, 'apps/nginx-sso-web-tls.conf.template'), path.join(ROOT, 'frontend/apps/sso-web/dist'),
      { APP_PREFIX: 'sso-app', APP_ORIGIN: origins.sso, BACKEND_SERVER1: backend, BACKEND_SERVER2: backend },
      { port: ports.sso, tlsDir: 'sso-web' });

    const database = `namewta_sso_test_${crypto.randomBytes(6).toString('hex')}`;
    const mysql = container('mysql', ['-p', '127.0.0.1::3306', '-e', 'MYSQL_ROOT_PASSWORD=owned-sso-test-only',
      '-e', `MYSQL_DATABASE=${database}`, IMAGES.mysql, '--character-set-server=utf8mb4', '--collation-server=utf8mb4_general_ci']);
    await ready(['docker', 'exec', mysql, 'mysqladmin', 'ping', '-h', '127.0.0.1', '-uroot', '-powned-sso-test-only', '--silent'], 'alive');
    const mysqlPort = output(['docker', 'port', mysql, '3306/tcp']).split(':').at(-1);
    const redis = container('redis', ['-p', '127.0.0.1::6379', IMAGES.redis, 'redis-server', '--save', '', '--appendonly', 'no']);
    await ready(['docker', 'exec', redis, 'redis-cli', 'ping'], 'PONG');
    const redisPort = output(['docker', 'port', redis, '6379/tcp']).split(':').at(-1);
    const command = ['./mvnw', '-B', '-ntp', '-pl', 'wta-admin', '-am', 'test', '-Dtest=SsoHttpsSessionIntegrationTest,AuthClientContextSsoUnitTest',
      '-Dsurefire.failIfNoSpecifiedTests=false', '-Dsso.release.integration=true',
      `-Dsso.release.backend.host=${gateway}`, `-Dsso.release.backend.port=${backendPort}`,
      `-Dsso.redis.integration.port=${redisPort}`,
      `-Dsso.mysql.integration.url=jdbc:mysql://127.0.0.1:${mysqlPort}/${database}?useSSL=false&allowPublicKeyRetrieval=true&serverTimezone=Asia/Shanghai`,
      `-Dnamewta.repo.root=${ROOT}`, `-Dnamewta.sql.root=${path.join(ROOT, 'release-artifacts/docker/infrastructure/mysql/init')}`,
      ...Object.entries(origins).map(([app, origin]) => `-Dsso.release.${app}.origin=${origin}`)];
    environment.MAVEN_OPTS = `${environment.MAVEN_OPTS ?? ''} -Daether.connector.requestTimeout=60000 -Daether.connector.connectTimeout=10000 -Daether.connector.basic.threads=1 -Dmaven.artifact.threads=1`;
    await run(command, path.join(ROOT, 'backend'), environment);
    const attributes = (file) => {
      const xml = fs.readFileSync(file, 'utf8');
      const start = xml.match(/<testsuite\b([^>]*)\/?>/);
      const values = {};
      for (const match of start[1].matchAll(/([A-Za-z_:][\w:.-]*)\s*=\s*"([^"]*)"/g)) values[match[1]] = match[2];
      return Object.fromEntries(['tests', 'failures', 'errors', 'skipped'].map((key) => [key, Number(values[key])]));
    };
    record.tests = attributes(path.join(ROOT, 'backend/wta-admin/target/surefire-reports/TEST-org.namewta.test.sso.SsoHttpsSessionIntegrationTest.xml'));
    assertSame(record.tests, { tests: 1, failures: 0, errors: 0, skipped: 0 });
    record.client_context_tests = attributes(path.join(ROOT, 'backend/wta-admin/target/surefire-reports/TEST-org.namewta.web.controller.AuthClientContextSsoUnitTest.xml'));
    assertSame(record.client_context_tests, { tests: 3, failures: 0, errors: 0, skipped: 0 });
    exitCode = 0;
  } catch (failure) {
    record.failure = String(failure?.message ?? failure);
    console.log(record.failure);
  } finally {
    if (temporary) fs.rmSync(temporary, { recursive: true, force: true });
    for (const row of [...record.containers].reverse()) {
      if (row.name !== 'mysql' && row.name !== 'redis') {
        const logs = spawnSync('docker', ['logs', row.id], { encoding: 'utf8' });
        const safe = `${logs.stdout ?? ''}${logs.stderr ?? ''}`.replace(/\?[^\s"']+/g, '?[REDACTED]');
        fs.writeFileSync(sibling(`${row.name}.log`), safe);
      }
      const removed = spawnSync('docker', ['rm', '-fv', row.id], { encoding: 'utf8' });
      record.cleanup.push({ id: row.id, exit_code: removed.status ?? 1 });
      if (removed.status) exitCode = 1;
    }
    if (record.network) {
      const removed = spawnSync('docker', ['network', 'rm', record.network], { encoding: 'utf8' });
      record.cleanup.push({ network: record.network, exit_code: removed.status ?? 1 });
      if (removed.status) exitCode = 1;
    }
    record.after = inventory();
    record.resources_restored = JSON.stringify(record.before) === JSON.stringify(record.after);
    if (!record.resources_restored) exitCode = 1;
    record.exit_code = exitCode;
    save();
  }
  return exitCode;
}

const scriptPath = fs.realpathSync(fileURLToPath(import.meta.url));
if (process.argv[1] && fs.existsSync(process.argv[1]) && fs.realpathSync(process.argv[1]) === scriptPath) {
  main().then((code) => process.exit(code ?? 0)).catch((error) => {
    console.error(error.usage ? error.message : (error.stack || error.message));
    process.exit(error.usage ? 2 : 1);
  });
}
